geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/FileBasedMetadata.scala

```diff
@@ -328,37 +340,65 @@ class FileBasedMetadata(
     }
   }
 
+  /**
+   * Delete a seq of paths
+   *
+   * @param paths paths to delete
+   * @param threads number of threads to use
+   */
+  private def delete(paths: Iterable[Path], threads: Int): Unit = {
+    if (threads < 2) {
+      paths.foreach(fc.delete(_, false))
+    } else {
+      val ec = ExecutionContext.fromExecutorService(new CachedThreadPool(threads))
+      try {
+        val parPaths = paths.par
+        parPaths.tasksupport = new ExecutionContextTaskSupport(ec)
+        parPaths.foreach(fc.delete(_, false))
+      } finally {
+        ec.shutdown()
+      }
+    }
+  }
+
   private class DirectoryWorker(
       es: ExecutorService,
       phaser: Phaser,
-      dir: Path,
+      listDirectory: => RemoteIterator[FileStatus],
       result: ConcurrentHashMap[String, PartitionFiles]
     ) extends Runnable {
 
     override def run(): Unit = {
       try {
-        val iter = fc.listStatus(dir)
-        while (iter.hasNext) {
+        var i = phaser.getRegisteredParties + 1
+        val iter = listDirectory
+        while (iter.hasNext && i < PhaserUtils.MaxParties) {
           val status = iter.next
           val path = status.getPath
           lazy val name = path.getName
           if (status.isDirectory) {
-            phaser.register() // register the new worker thread
-            es.submit(new DirectoryWorker(es, phaser, path, result))
+            i += 1
+            // use a tiered phaser on each directory avoid the limit of 65535 registered parties
+            es.submit(new DirectoryWorker(es, new Phaser(phaser, 1), fc.listStatus(path), result))
           } else if (name.startsWith(UpdatePartitionPrefix)) {
             // pull out the partition name but don't parse the file yet
             val encoded = name.substring(8, name.length - 42) // strip out prefix and suffix
             val partition = StringSerialization.decodeAlphaNumericSafeString(encoded)
             result.merge(partition, PartitionFiles(unparsed = Seq(path)), addFiles)
           } else if (name == CompactedHocon || name == CompactedJson) {
+            i += 1
             phaser.register() // register the new worker thread
             es.submit(new CompactedParser(phaser, path, result))
           } else if (name.startsWith(UpdateFilePrefix) && name.endsWith(JsonPathSuffix)) {
             // old update files - have to parse them to get the partition name
+            i += 1
             phaser.register() // register the new worker thread
             es.submit(new UpdateParser(phaser, path, result))
           }
         }
+        if (iter.hasNext) {
```

Comment:
Ok, if I'm understanding the strategy, as the iterator is looped through, tasks are submitted up to the 65K limit for the Phaser to manage.  

How quickly does this happen?  Basically, could there be a situation where the JVM heap has millions of objects on it representing the work that needs to be done and that causes an OOM?

Reply:
ah, potentially... but these objects are fairly lightweight, all the state is mostly shared among workers (the iterator of results, the concurrent map to return results). you'd end up with at most a task per file or directory (and briefly some duplicates when/if it needs to pass off to a new phaser to list out a whole directory), so it's limited by the number of files you have on disk and how fast it parses through them. in practice it definitely consumes a decent amount of memory overall when you have a lot of files, but likely a lot of that is just keeping the data for each file in the cache.

Reply:
i think we'd have to re-architect things a bit for that to work. with a finite queue you'd have to use a different thread pool for the producer and the consumer, otherwise your threads could all be used up by producers waiting to put things in the queue, with no consumers able to read things back off. we'd need to come up with a way to be notified when all the threads are done, as well. yeah, it might be worthwhile to re-think this approach at some point, and that might be a good replacement.

Reply:
Cool.